win32/devdoc/srw_lock_ll_win32_requirements.md

```diff
@@ -0,0 +1,127 @@
+# `srw_lock_ll_win32` requirements
+
+## Overview
+
+`srw_lock_ll_win32` is the implementation of a slim reader writer lock (`srw_lock_ll` interface). On Windows it simply wraps `SRWLOCK`.
+
+## Exposed API
+
+```c
+#define SRW_LOCK_LL_TRY_ACQUIRE_RESULT_VALUES \
+    SRW_LOCK_LL_TRY_ACQUIRE_OK, \
+    SRW_LOCK_LL_TRY_ACQUIRE_COULD_NOT_ACQUIRE, \
+    SRW_LOCK_LL_TRY_ACQUIRE_INVALID_ARGS
+
+typedef void* SRW_LOCK_LL;
+
+MU_DEFINE_ENUM(SRW_LOCK_LL_TRY_ACQUIRE_RESULT, SRW_LOCK_LL_TRY_ACQUIRE_RESULT_VALUES)
+
+MOCKABLE_FUNCTION(, void, srw_lock_ll_init);
+MOCKABLE_FUNCTION(, void, srw_lock_ll_deinit, SRW_LOCK_LL*, srw_lock_ll);
+
+/*writer APIs*/
+MOCKABLE_FUNCTION(, void, srw_lock_ll_acquire_exclusive, SRW_LOCK_LL*, srw_lock_ll);
+MOCKABLE_FUNCTION(, SRW_LOCK_LL_TRY_ACQUIRE_RESULT, srw_lock_ll_try_acquire_exclusive, SRW_LOCK_LL*, srw_lock_ll);
+MOCKABLE_FUNCTION(, void, srw_lock_ll_release_exclusive, SRW_LOCK_LL*, srw_lock_ll);
+
+/*reader APIs*/
+MOCKABLE_FUNCTION(, void, srw_lock_ll_acquire_shared, SRW_LOCK_LL*, srw_lock_ll);
+MOCKABLE_FUNCTION(, SRW_LOCK_LL_TRY_ACQUIRE_RESULT, srw_lock_ll_try_acquire_shared, SRW_LOCK_LL*, srw_lock_ll);
+MOCKABLE_FUNCTION(, void, srw_lock_ll_release_shared, SRW_LOCK_LL*, srw_lock_ll);
+```
+
+### srw_lock_ll_init
+```c
+MOCKABLE_FUNCTION(, void, srw_lock_ll_init, SRW_LOCK_LL*, srw_lock_ll);
+```
+
+`srw_lock_ll_init` initializes a slim reader writer lock.
+
+If `srw_lock_ll` is `NULL`, `srw_lock_ll_init` shall return.
+
+Otherwise, `srw_lock_ll_init` shall call `InitializeSRWLock`.
+
+### srw_lock_ll_deinit
+```c
+MOCKABLE_FUNCTION(, void, srw_lock_ll_deinit, SRW_LOCK_LL*, srw_lock_ll);
+```
+
+`srw_lock_ll_deinit` frees deinitializes the slim reader writer lock.
+
+If `srw_lock_ll` is `NULL` then `srw_lock_ll_deinit` shall return.
+
+Otherwise, `srw_lock_ll_deinit` shall return.
+
+### srw_lock_ll_acquire_exclusive
+```c
+MOCKABLE_FUNCTION(, void, srw_lock_ll_acquire_exclusive, SRW_LOCK_LL*, srw_lock_ll);
+```
+
+`srw_lock_ll_acquire_exclusive` acquires the slim reader writer lock in exclusive (writer) mode.
+
+If `srw_lock_ll` is `NULL` then `srw_lock_ll_acquire_exclusive` shall return.
```

Comment:
>shall return [](http://example.com/codeflow?start=64&length=12)

I do have to question here the decision to "return". Upon returning, the calling thread will rightfully believe that it have ownership over some data, while in reality, it does not.

What would happen if it would eb specced as "shall not return"?

Reply:
After some tinkering I looked at what we use for Linux and decided will make this return an int as the API we call for pthreads can fail.

And the caller can decide whether to ignore or not this error but it will have the information.